scripts/builtin/hmm.dml

```diff
@@ -0,0 +1,181 @@
+{
+    /*
+    alpha a matrix of size nr_states * T with a cell i,t being probability of
+    the model producing outputs (o_t+1,..., o_T) given that the model is 
+    at state i at time t
+    */
+    
+    T = col(X)
+    nr_states = row(A)
+    beta = matrix(0, rows=nr_states, cols=T)
+    beta[,T] = matrix(1, rows=length(ip), cols=1)
+    
+    for (t in (T-1):1) {
+        for (i in 1:nr_states) {
+            beta[i, t] = sum(beta[, t+1] * A[i, ] * B[ , X[t]])
+        }
+    }
+}
+
+calculate_gamma = function (Matrix[Double] alpha, Matrix[Double] beta) return (Matrix[Double] gamma)
+{
+    /*
+    gamma a nr_state * T matrix with cell (i, t) being probability of 
+    the state being at i at timestep j given the observed output 
+    */
+    nr_states = nrow(alpha)
+    T = ncol(alpha)
+    gamma = matrix(1/nr
+    parfor (i in 1:nrow(alpha)) {
+        for (t in 1:ncol(alpha)) {
+            num_ij = alpha[i, t] * beta[i, t]
+            den_ij = sum(alpha[,t] * beta[,t]
+            gamma[i, j] = num_ij / den_ij
+    }
+   }        
+}
+
+calculate_eta =  function (Matrix[Double] alpha, Matrix[Double] beta, Matrix[Double] A, Matrix[Double] B) return (Matrix[Double] eta)
+{
+    /*
+    gamma a (nr_states * nr_states) * T matrix with cell (i, t) being probability of 
+    the state being at i at timestep j given the observed output
+    */
+    nr_states = nrow(alpha)
+    T = ncol(alpha)
+    tot_transitions = nr_states * nr_states
+    eta = matrix(1/nr_states, rows=tot_transitions, cols=T-1)
+    
+    /*
+    The transitions will be indiced as such:transition 1-N will represent
+    transition from state 1 to 1, 2 upto state N. transition N+1-2N will represent
+    transitions from 2 to 1, 2 upto state N  
+    */
+    
+    parfor (trans_id in 1:tot_transitions) {
+        for (t in 1:(T-1)) {
+            #indices for alpha and beta
+            i = floor(trans_id / nr_states) + 1 #index starts at 1
+            j = trans_id - ((i-1) * nr_states)
+            num_ij = alpha[i, t] * A[i, j] * beta[j, t+1] * B[j, X[t+1]]
+            den_ij = sum(sum(alpha[, t]))
+            eta[trans_id, t] = alpha[i, j]
+        }
+    }
+}
+
+
+baum_welch = function (Matrix[Double] X, nr_states, nr_outputs) return (Matrix[Double] A, Matrix[Double] B, Matrix[Double] ip, likelihood)
+{
+    #initialize state transition and emmission matrices uniformly
+    A = matrix(1/nr_states, rows=nr_states, cols=nr_states)
+    B = matrix(1/nr_outputs, rows=nr_states, cols=nr_ouptuts)
+    ip = matrix(1/nr_states, rows=nr_states, cols=1)
+    
+    converge = FALSE
+    
+    while (!converge) {
+        alpha = forward(X, A, B, ip)
+        beta = backward(X, A, B)
+        
+        gamma = calculate_gamma(alpha)
+        eta = calculate_eta(alpha, A, B)       
+        /*
+        TODO: compute likelihood, if it does not change much from previous
+        iteration, break.
+        */
+    }
+}
```

Comment:
new lines in end of files.

Reply:
Resolved

Reply:
I do not understand this. Could you please clarify?

Reply:
simple, GitHub indicate with a big red mark that there is no newline. The code does work, but we try to make it consistent across files.

Reply:
Should I add or remove a newline?

Reply:
add